src/io/writer_internal.cpp

```diff
@@ -0,0 +1,209 @@
+    // Set the WKT bit.
+    head14.global_encoding |= (1 << 4);
+
+    out_stream.seekp(0);
+    head14.write(out_stream);
+
+    // Write the VLR.
+    copc_data.span = file->GetCopc().span;
+    copc_data.header().write(out_stream);
+    copc_data.write(out_stream);
+
+    lazVlr.header().write(out_stream);
+    lazVlr.write(out_stream);
+
+    if (head14.ebCount())
+    {
+        auto ebVlr = this->file->GetExtraBytes();
+        ebVlr.header().write(out_stream);
+        ebVlr.write(out_stream);
+    }
+}
+
+void WriterInternal::WriteWkt(las::LasHeader &head14)
+{
+    auto wkt = file->GetWkt();
+    evlr_header h{0, "LASF_Projection", 2112, (uint64_t)wkt.size(), ""};
+
+    wkt_vlr vlr(wkt);
+
+    out_stream.seekp(0, std::ios::end);
+    int64_t offset = static_cast<int64_t>(out_stream.tellp());
+    copc_data.wkt_vlr_offset = offset + evlr_header::Size;
+    copc_data.wkt_vlr_size = vlr.size();
+
+    h.write(out_stream);
+    vlr.write(out_stream);
+
+    // TODO: Let the Page writer set the evlr offset if no wkt exists
+    head14.evlr_offset = offset;
+    head14.evlr_count++;
+}
+
+void WriterInternal::WriteChunkTable()
+{
+    // move to the end of the file to start emitting our compresed table
+    out_stream.seekp(0, std::ios::end);
+
+    // take note of where we're writing the chunk table, we need this later
+    int64_t chunk_table_offset = static_cast<int64_t>(out_stream.tellp());
+
+    // Fixup the chunk table to be relative offsets rather than absolute ones.
+    uint64_t prevOffset = FIRST_CHUNK_OFFSET();
+    for (auto &c : chunks_)
+    {
+        uint64_t relOffset = c.offset - prevOffset;
+        prevOffset = c.offset;
+        c.offset = relOffset;
+    }
+
+    // write out the chunk table header (version and total chunks)
+    uint32_t version = 0;
+    out_stream.write((const char *)&version, sizeof(uint32_t));
+    uint32_t numChunks = (uint32_t)chunks_.size();
+    out_stream.write((const char *)&numChunks, sizeof(uint32_t));
+
+    // Write the chunk table
+    OutFileStream w(out_stream);
+
+    compress_chunk_table(w.cb(), chunks_, true);
+    // go back to where we're supposed to write chunk table offset
+    out_stream.seekp(OFFSET_TO_POINT_DATA);
+    out_stream.write(reinterpret_cast<char *>(&chunk_table_offset), sizeof(chunk_table_offset));
+}
+
+// Writes a node and returns the node's offset and size in the file
+Entry WriterInternal::WriteNode(std::vector<char> in, uint64_t point_count, bool compressed)
+{
+    Entry entry;
+    entry.offset = out_stream.tellp();
+
+    if (compressed)
+        out_stream.write(in.data(), in.size());
+    else
+        point_count = laz::Compressor::CompressBytes(out_stream, file->GetLasHeader(), in);
+
+    point_count_14_ += point_count;
+
+    uint64_t endpos = (uint64_t)out_stream.tellp();
+    chunks_.push_back(lazperf::chunk{point_count, endpos});
+
+    entry.size = endpos - entry.offset;
+    entry.point_count = point_count;
```

Comment:
`entry.point_count` is int32 when `point_count` is uint64, maybe we should check and throw error if value exceed?